builtin/rebase.c

```diff
@@ -0,0 +1,650 @@
+		strbuf_addf(&buf, "%s/rewritten", merge_dir());
+		if (is_directory(buf.buf)) {
+			options.type = REBASE_PRESERVE_MERGES;
+			options.flags |= REBASE_INTERACTIVE_EXPLICIT;
+		} else {
+			strbuf_reset(&buf);
+			strbuf_addf(&buf, "%s/interactive", merge_dir());
+			if(file_exists(buf.buf)) {
+				options.type = REBASE_INTERACTIVE;
+				options.flags |= REBASE_INTERACTIVE_EXPLICIT;
+			} else
+				options.type = REBASE_MERGE;
+		}
+		options.state_dir = merge_dir();
+	}
+
+	if (options.type != REBASE_UNSPECIFIED)
+		in_progress = 1;
+
+	total_argc = argc;
+	argc = parse_options(argc, argv, prefix,
+			     builtin_rebase_options,
+			     builtin_rebase_usage,
+			     PARSE_OPT_STOP_AT_NON_OPTION);
+
+	if (action != NO_ACTION && total_argc != 2) {
+		usage_with_options(builtin_rebase_usage,
+				   builtin_rebase_options);
+	}
+
+	if (argc > 2)
+		usage_with_options(builtin_rebase_usage,
+				   builtin_rebase_options);
+
+	if (!(options.flags & REBASE_VERBOSE))
+		strbuf_addstr(&git_am_opt, " -q");
+	options.git_am_opt = git_am_opt.buf;
+
+	if (action != NO_ACTION && !in_progress)
+		die(_("No rebase in progress?"));
+
+	switch (action) {
+	case ACTION_CONTINUE: {
+		struct object_id head;
+		struct lock_file lock_file = LOCK_INIT;
+		int fd;
+
+		options.action = "continue";
+
+		/* Sanity check */
+		if (get_oid("HEAD", &head))
+			die(_("Cannot read HEAD"));
+
+		fd = hold_locked_index(&lock_file, 0);
+		refresh_index(the_repository->index, REFRESH_QUIET, NULL, NULL,
+			      NULL);
+		if (0 <= fd)
+			update_index_if_able(the_repository->index,
+					     &lock_file);
+		rollback_lock_file(&lock_file);
+
+		if (has_unstaged_changes(1)) {
+			puts(_("You must edit all merge conflicts and then "
+			       "mark them as resolved using git add"));
+			exit(1);
+		}
+		if (read_basic_state(&options))
+			exit(1);
+		goto run_rebase;
+	}
+	case ACTION_SKIP:
+		options.action = "skip";
+
+		if (reset_hard(NULL) < 0)
+			die(_("could not discard worktree changes"));
+		if (read_basic_state(&options))
+			exit(1);
+		goto run_rebase;
+	case ACTION_ABORT: {
+		struct string_list merge_rr = STRING_LIST_INIT_DUP;
+		options.action = "abort";
+
+		rerere_clear(&merge_rr);
+		string_list_clear(&merge_rr, 1);
+
+		if (read_basic_state(&options))
+			exit(1);
+		if (starts_with(options.head_name, "refs/") &&
+		    create_symref("HEAD", options.head_name,
+				  "rebase: aborting") < 0)
+			die(_("could not move back to %s"), options.head_name);
+		if (reset_hard(&options.orig_head) < 0)
+			die(_("could not move back to %s"),
+			    oid_to_hex(&options.orig_head));
+		ret = finish_rebase(&options);
+		goto cleanup;
+	}
+	case ACTION_QUIT: {
+		strbuf_reset(&buf);
+		strbuf_addstr(&buf, options.state_dir);
+		ret = !!remove_dir_recursively(&buf, 0);
+		if (ret)
+			die(_("could not remove '%s'"), options.state_dir);
+		goto cleanup;
+	}
+	case ACTION_EDIT_TODO:
+	case ACTION_SHOW_CURRENT_PATCH:
+		goto run_rebase;
+	default:
+		die("TODO");
```

Comment:
This is wrong, and breaks the `git rebase <upstream>` case: the value of `action` is `NO_ACTION` in this case, and therefore we fall into this `"TODO"` trap.

I think this diff should fix that:

```diff
@@ -546,8 +559,10 @@ int cmd_rebase(int argc, const char **argv, const char *prefix)
        case ACTION_EDIT_TODO:
        case ACTION_SHOW_CURRENT_PATCH:
                goto run_rebase;
+       case NO_ACTION:
+               break;
        default:
-               die("TODO");
+               BUG("action: %d", action);
        }

        options.type = REBASE_AM;
```